### ISSUE TITLE
Support type hinting for variables

### ISSUE THREAD

**Issue by belazor** [felixfbecker/php-language-server#350]
If I have a variable that's type hinted like so:

<img width="830" alt="screenshot 2017-04-11 at 8 41 54 pm" src="https://cloud.githubusercontent.com/assets/9424286/24927586/6ce33d82-1ef7-11e7-8a66-eb9d98ee8d18.png">

I expect to see the method definitions for the methods I call on the $emailChange variable. This does not happen:

<img width="831" alt="screenshot 2017-04-11 at 8 43 42 pm" src="https://cloud.githubusercontent.com/assets/9424286/24927640/a301e882-1ef7-11e7-93be-6926fd6cc8cf.png">

This type hinting syntax is allegedly supported by phpStorm, and it would be wonderful to have support for this in VSCode via your IntelliSense plugin as well.

Thank you in advance :)

**Comment by felixfbecker** [comment]
Definitely, there is already handling for `@params` tags, so it would be trivial to add `@var` tags handling if someone wants to take a shot at it.

**Comment by belazor** [comment]
Would you be able to give me a hint as to where to implement this? I could probably handle it if I had an idea of where to look :)

**Comment by felixfbecker** [comment]
In this function: https://github.com/felixfbecker/php-language-server/blob/master/src/DefinitionResolver.php#L726

**Comment by belazor** [comment]
I guess I'm not advanced enough a PHP coder, I can't work out how to actually test any changes I make. Trying to run the web server just produces a bunch of Undefined Index errors related to headers in the ProtocolStreamer.php file.

I'm seeing around line 798 that variables should have @var tags supported, unless I'm just horribly misreading the code? Does that make the fact that this isn't working a bug?

If it isn't a bug and I'm just failing at reading your code, would you please consider adding it yourself, as you're more intimately familiar with the code? I would really appreciate it!

Sorry I couldn't be of more help :(

**Comment by felixfbecker** [comment]
The language server has nothing to do with a web server or headers ;)
See the README for how to run tests.
`@var` tags are read for properties, not variables.

**Comment by belazor** [comment]
I still don't understand, sorry. 

The only mention of testing that I could find in the README file was a reference to unit testing. I don't have much experience with unit testing, so me running a command that says "149 / 149 (100%)" doesn't do much when I'm struggling with actually working out how to read the stubs file to find out if any changes are working. 

I guess I just have to hope someone is kind enough to implement this for me :)

**Comment by jens1o** [comment]
Well, then I take a look and check wether I can do that.

**Comment by jens1o** [comment]
ah... thanks. Didn't expected that.

**Comment by felixfbecker** [comment]
Is this still an issue in the latest version?

**Comment by belazor** [comment]
@felixfbecker I have tested this in v1.4.1 and it is still an issue.

```php
		/** @var \DBTech\Credits\Entity\Currency $currency */
		if (!$currency = DBTech::em()->findOne('DBTech\Credits:Currency', $cleanedInput['id']))
		{
			// Bad currency
			throw $this->errorException(DBTech::phrase('dbtech_credits_invalid_currency'));
		}

		if (!$currency->isActive())
		{
			// Bad currency
			throw $this->errorException(DBTech::phrase('dbtech_credits_invalid_currency'));
		}
```
I receive no type hinting for **isActive**.

That method of docblocking allegedly works in phpStorm (untested, but this information comes from the XenForo product developers who use phpStorm internally) but does not appear to work here.

**Comment by felixfbecker** [comment]
Does it work if you write it like 
```php
/** @var \DBTech\Credits\Entity\Currency $currency */
$currency = DBTech::em()->findOne('DBTech\Credits:Currency', $cleanedInput['id']);
```

**Comment by belazor** [comment]
Small sidenote, as I was cutting the variable definition out of that line I appear to be triggering a bug.

https://i.imgur.com/wacZGc9.png

It is not recognising that I typed an expression, and also the **php** process is running rampant, using 100% CPU and causing my fans to go into overdrive. The only fix is to completely close VSCode and Force Quit the php process, then restarting VSCode.

This happens seemingly randomly, and has happened since I first installed this extension around the time I opened this issue.

Is this a known issue?

Regardless;

```php
		/** @var \DBTech\Credits\Entity\Currency $currency */
		$currency = DBTech::em()->findOne('DBTech\Credits:Currency', $cleanedInput['id']);

		if (!$currency)
		{
			// Bad currency
			throw $this->errorException(DBTech::phrase('dbtech_credits_invalid_currency'));
		}

		if (!$currency->isActive())
		{
			// Bad currency
			throw $this->errorException(DBTech::phrase('dbtech_credits_invalid_currency'));
		}
```
Still produces "No definition found for isActive".

**Comment by felixfbecker** [comment]
Could you file that as another issue?

**Comment by belazor** [comment]
No problem, I was just checking to see if it was a known issue so I didn't clog up the system with a dupe, I'll go ahead and do that right now :)

**Comment by jens1o** [comment]
This does not work completly. Indeed it will be shown, but there are no suggestions for a type hinted var.
For example: 
![image](https://user-images.githubusercontent.com/11234139/27540487-a3829828-5a80-11e7-8760-c6153b0f8701.png)


While I have no suggestions:
![image](https://user-images.githubusercontent.com/11234139/27540495-aa0a4f88-5a80-11e7-820a-b4f8512c87bf.png)


When I create a new instance of it and make it explicit, it works:
![image](https://user-images.githubusercontent.com/11234139/27540466-95443294-5a80-11e7-85c5-49d62413a405.png)

I'm using 
https://docs.planetteamspeak.com/ts3/php/framework/index.html

**Comment by felixfbecker** [comment]
yes, that's why this issue is open

**Comment by jens1o** [comment]
Ah, I thought it's not even implemented. I was kinda suprised why this issue exists, when it exists, but there is no handling... Hmm.

**Comment by nguyentamgm** [comment]
~~Oh, exactly what I am looking for, hopefully this will be implemented soon! Even this issue was here for over a year.~~

Sorry, it works if I add 1 more *, just a little bit difference from PHPStorm
`/* @var File $file */` => `/** @var File $file */`

**Comment by vlad88sv** [comment]
Thanks for adding this feature, any chance to also support Netbeans/Eclipse/Zend style?
[More info in this link](https://stackoverflow.com/a/22666308/1172180)

The reason behind this request is not having to convert hundreds of files of legacy code so they work with VSCode.

Example of Netbeans/Eclipse/Zend style:
/* @var $varName Type_Name */

**Comment by khooz** [comment]
Hi,

Making some traits here... `$this` type-hinting for traits will be a tremendous help.

Cheers!

**Comment by twinklebob** [comment]
So are we just waiting for PR #701 to have a unit test or two added and then this feature is ready to go?

**Comment by vinkla** [comment]
I know this is off topic but is this already possible with JavaScript in Visual Studio Code?

**Comment by iddqd3** [comment]
> ~Oh, exactly what I am looking for, hopefully this will be implemented soon! Even this issue was here for over a year.~
> 
> Sorry, it works if I add 1 more *, just a little bit difference from PHPStorm
> `/* @var File $file */` => `/** @var File $file */`

This does not work as expected. Errors disappeared but the transition to the definition does not work, and phpunit also gives an error

`/** @var Repository\RequestRepository $requestsRepo */
        $requestsRepo = $this->em->getRepository (Request::class);
        $requests = $requestsRepo->getRequests ($entity);`


`115 Call to method getRequests() on an unknown class ..\Model\Repository\RequestRepository.`

**Comment by shushenghong** [comment]
> Thanks for adding this feature, any chance to also support Netbeans/Eclipse/Zend style?
> [More info in this link](https://stackoverflow.com/a/22666308/1172180)
> 
> The reason behind this request is not having to convert hundreds of files of legacy code so they work with VSCode.
> 
> Example of Netbeans/Eclipse/Zend style:
> /* @var $varName Type_Name */

need support :)

**Comment by julesgilson** [comment]
so that PR #701 is still sitting there - 4 years now? Or did some other change make variable type hinting work (I can't get it to)

**Comment by Ali-restaurant** [comment]
> In this function: https://github.com/felixfbecker/php-language-server/blob/master/src/DefinitionResolver.php#L726





https://phps.space/questions/issues-encountered-with-php-variables-on-a-different-server
